File: rfcs/bench-0104/harness/c/pidigits.c
/* pidigits, from the Computer Language Benchmarks Game.
 *
 * A transcription of examples/pidigits.vyrn. Build with:
 *
 *   clang -O2 -ffp-contract=off -std=c11 -o pidigits.exe pidigits.c
 *
 * N is argv[1]; without it the census N is used.
 *
 * This is the bounded spigot of Rabinowitz and Wagon, the same one the Vyrn
 * program runs -- it holds one small integer per digit column, so 64-bit
 * arithmetic is enough and no big integer is needed.
 */

#include <stdio.h>
#include <stdlib.h>

/* The census N -- 27 digits. */
#define CENSUS_ORDER 27

/* Extra digits computed and thrown away. A bounded spigot's last few columns
 * are the ones that can be wrong, so the bound is set past the answer. */
#define GUARD 10

static void *checkedMalloc(size_t bytes)
{
    void *p = malloc(bytes);

    if (p == NULL) {
        fprintf(stderr, "out of memory\n");
        exit(1);
    }
    return p;
}

/* The first `n` digits of pi, one per entry. */
static long long *piDigits(long long n)
{
    long long total = n + GUARD;
    long long len = 10 * total / 3 + 1;
    long long *a = (long long *)checkedMalloc((size_t)len * sizeof(long long));
    long long *out = (long long *)checkedMalloc((size_t)(total + 2) * sizeof(long long));
    long long *digits = (long long *)checkedMalloc((size_t)(n > 0 ? n : 1) * sizeof(long long));
    long long outLen = 0;
    long long nines = 0;
    long long predigit = 0;
    long long i, j, d;

    for (i = 0; i < len; i++) {
        a[i] = 2;
    }

    for (j = 0; j <= total; j++) {
        long long q = 0;
        long long k = len;
        while (k > 0) {
            long long x = 10 * a[k - 1] + q * k;
            a[k - 1] = x % (2 * k - 1);
            q = x / (2 * k - 1);
            k = k - 1;
        }
        a[0] = q % 10;
        q = q / 10;
        /* A run of nines is held back: the carry out of the next column can turn
         * all of them into zeroes and bump the digit before them. */
        if (q == 9) {
            nines = nines + 1;
        } else if (q == 10) {
            long long z;
            out[outLen++] = predigit + 1;
            for (z = 0; z < nines; z++) {
                out[outLen++] = 0;
            }
            predigit = 0;
            nines = 0;
        } else {
            long long z;
            out[outLen++] = predigit;
            predigit = q;
            for (z = 0; z < nines; z++) {
                out[outLen++] = 9;
            }
            nines = 0;
        }
    }

    /* Entry 0 is the zero that precedes the 3; the guard digits fall off the
     * end. */
    for (d = 1; d <= n; d++) {
        digits[d - 1] = out[d];
    }

    free(a);
    free(out);
    return digits;
}

/* The digits, ten to a line, each line tagged with how many have been printed.
 * A short last line is padded to ten so the tags stay in one column. */
static void run(long long n)
{
    long long *digits = piDigits(n);
    char line[16];
    int len = 0;
    long long i;

    for (i = 0; i < n; i++) {
        line[len++] = (char)('0' + digits[i]);
        if ((i + 1) % 10 == 0) {
            line[len] = '\0';
            printf("%s\t:%lld\n", line, i + 1);
            len = 0;
        }
    }
    if (len > 0) {
        while (len < 10) {
            line[len++] = ' ';
        }
        line[len] = '\0';
        printf("%s\t:%lld\n", line, n);
    }
    free(digits);
}

int main(int argc, char **argv)
{
    long long order = CENSUS_ORDER;

    if (argc > 1) {
        order = strtoll(argv[1], NULL, 10);
    }
    run(order);
    return 0;
}

File: rfcs/bench-0104/harness/rust/pidigits.rs
// pidigits, from the Computer Language Benchmarks Game.
//
// The Rust leg of RFC-0104 M2. It computes what `examples/pidigits.vyrn`
// computes and prints the same bytes.
//
//   $ rustc -C opt-level=3 -o pidigits.exe pidigits.rs
//   $ ./pidigits 27
//
// Safe Rust, std only — and, to stay the same program as the Vyrn one, the same
// bounded spigot of Rabinowitz and Wagon over `i64` rather than an arbitrary-
// precision entry. A big-integer entry is a different program; this row compares
// the two implementations of the same one.

use std::env;
use std::io::{self, BufWriter, Write};

/// The census N — 27 digits.
const ORDER: i64 = 27;

/// Extra digits computed and thrown away. A bounded spigot's last few columns are
/// the ones that can be wrong, so the bound is set past the answer.
const GUARD: i64 = 10;

/// The first `n` digits of pi, one per entry.
fn pi_digits(n: i64) -> Vec<i64> {
    let total = n + GUARD;
    let len = 10 * total / 3 + 1;
    let mut a = vec![2i64; len as usize];
    let mut out: Vec<i64> = Vec::new();
    let mut nines = 0;
    let mut predigit = 0;
    let mut j = 0;
    while j <= total {
        let mut q = 0;
        let mut k = len;
        while k > 0 {
            let x = 10 * a[(k - 1) as usize] + q * k;
            a[(k - 1) as usize] = x % (2 * k - 1);
            q = x / (2 * k - 1);
            k = k - 1;
        }
        a[0] = q % 10;
        q = q / 10;
        // A run of nines is held back: the carry out of the next column can turn
        // all of them into zeroes and bump the digit before them.
        if q == 9 {
            nines = nines + 1;
        } else if q == 10 {
            out.push(predigit + 1);
            for _ in 0..nines {
                out.push(0);
            }
            predigit = 0;
            nines = 0;
        } else {
            out.push(predigit);
            predigit = q;
            for _ in 0..nines {
                out.push(9);
            }
            nines = 0;
        }
        j = j + 1;
    }
    // Entry 0 is the zero that precedes the 3; the guard digits fall off the end.
    let mut digits = Vec::with_capacity(n as usize);
    let mut d = 1;
    while d <= n {
        digits.push(out[d as usize]);
        d = d + 1;
    }
    digits
}

/// The digits, ten to a line, each line tagged with how many have been printed.
/// A short last line is padded to ten so the tags stay in one column.
fn run(out: &mut impl Write, n: i64) {
    let digits = pi_digits(n);
    let mut line = String::new();
    let mut i = 0;
    while i < n {
        line.push_str(&digits[i as usize].to_string());
        if (i + 1) % 10 == 0 {
            writeln!(out, "{}\t:{}", line, i + 1).unwrap();
            line.clear();
        }
        i = i + 1;
    }
    if !line.is_empty() {
        while line.len() < 10 {
            line.push(' ');
        }
        writeln!(out, "{}\t:{}", line, n).unwrap();
    }
}

fn main() {
    let n: i64 = env::args()
        .nth(1)
        .map(|a| a.parse().expect("N must be an integer"))
        .unwrap_or(ORDER);

    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    run(&mut out, n);
}
